### ISSUE TITLE
PHP proxy fails to pass back full response

### ISSUE THREAD

**Issue by hgonzago** [Esri/resource-proxy#190]
This came about when trying to use the most recent PHP proxy against the basic directions Javascript sample. Swapping out the proxy file with the PHP one on my machine (OSX Mavericks, and then Yosemite) and adding my credentials to the config file. When running the same everytime i would get an "Undefined token error" and after checking the response back from the route service via the proxy, it was always truncated, "dojo.io.script.jsonp_dojoIoScript1._jsonpCallback({"currentVersion":10.3,"layerName":"Route_World","layerType":"esriNAServerRouteLayer","impedance":"TravelTime","restrictions":["Avoid Unpaved Roads","Avoid "

Notice that the full JSON callback is cut off. To narrow down if it was machine specific, I configured a clean Apache instance on my personal Macbook running Yosemite (older hardware model, via 2012) and got the same exact error message when trying to run this sample.

Spoke to @phpmaps and he had an idea what was going on. Created this issue so that the fix he implemented in an updated PHP could be merged into this repo.

**Comment by phpmaps** [comment]
The issue appears to be related to older GZIP concerns.  As PHP interrogates GZIP'ed resource responses (looking for 403 failures and errors) it decompresses the data.   I believe in some cases the proxy brokers the response equal to the Content-Length size (of the gzip'ed response) instead of calculating a new Content-Length size based on the decompressed data.  Since one goal is to proxy content according to the original requests and responses - my plan is to clone the ArcGIS Online and Server response and interrogate the other.  If all is good - send the client the cloned response.  Repackaging would be expensive and more code - plus requires the zlib extension which not everyone may have.

http://php.net/manual/en/function.gzcompress.php

**Comment by hgonzago** [comment]
@phpmaps - Awesome, thanks so much for digging into this and figuring out the issue.:-)

**Comment by phpmaps** [comment]
See #193

**Comment by driskull** [comment]
I'm running into this one as well with PHP.

**Comment by phpmaps** [comment]
@driskull  Have you tried this new proxy.php file awaiting a merge?  I am very hopeful that this fixes the issue.  I suspect that this one line will make a big difference in preventing truncation.

https://github.com/phpmaps/resource-proxy/blob/patch-1/PHP/proxy.php#L576-578

**Comment by driskull** [comment]
Yes I did and it does work.

**Comment by driskull** [comment]
Actually, it doesn't work for me. It did for a while but now it isn't.

**Comment by phpmaps** [comment]
@driskull  A couple of questions.  Are you testing with this source here https://github.com/phpmaps/resource-proxy/tree/patch-1 ?  Second, what type of server are you using when testing?  Lastly, if you see the problem again, can you post the HTTP Response headers in this thread.

The good news is @bsvensson has yet to merge this in b/ he say duplicate headers which I think were specific to a Windows IIS or Windows Apache environment .  I would make sure you are using this source.

https://github.com/phpmaps/resource-proxy/tree/patch-1

**Comment by hgonzago** [comment]
@phpmaps Not sure if the duplicate headers were specific to Windows IIS or Windows APache. Pretty sure I saw it on my end using Apache on my Mac OS.

**Comment by driskull** [comment]
It seems to be working now. I'll see if i can reproduce it again.

**Comment by bsvensson** [comment]
@driskull - make sure it wasn't an issue due to expired-token not being renewed - #146.

**Comment by driskull** [comment]
I ran into the expired token on .Net version. PHP seems to be working but i'll keep an eye on it.

**Comment by hgonzago** [comment]
I'm seeing some weird stuff with this @phpmaps @bsvensson  - 

if using a JSAPI app and have it set to esriConfig.defaults.io.alwaysUseProxy = true; instead of urlUtils.addProxyRule({
          urlPrefix: ...

I end up getting the proxy duplicated in the request URL header and therefore get a 403 forbidden error saying it can't access this.

eg.
http://heatherg.esri.com/proxy/proxy.php?http://heatherg.esri.com/proxy/proxy.php?http://services.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer/tilemap/18/109472/71656/8/8

Another weird thing is that the the services.arcgisonline.com URL is set in the proxy config as well. It only bails on the tiled layers.

**Comment by hgonzago** [comment]
@phpmaps I'm dreading this but it looks like this problem reared it's nasty head again.

I'm on a brand new machine. Just got it today so no crazy configuration settings or anything. Using default PHP and apache that comes with the Mac OSX. Took the latest version you currently have up and tried using this to proxy app credentials to run the directions sample located here, 

everytime I try to run it, I always get the truncated response similar to what I had the first time this popped up, 

"dojo.io.script.jsonp_dojoIoScript1._jsonpCallback({"currentVersion":10.3,"layerName":"Route_World","layerType":"esriNAServerRouteLayer","impedance":"TravelTime","restrictions":["Avoid Unpaved Roads","Avoid Private Roads","Driv

**Comment by phpmaps** [comment]
Ok.  Good to know and disappointing to hear.  Would you say this issue is intermitent?  Also just to confirm are you still using the code in this pull request and not the master?  Also if it happens again can you post response headers?

**Comment by hgonzago** [comment]
I just now realized I think I did a copy and paste from the master for something else I noticed last week so wanted to make sure I was looking at the right stuff. I just grabbed the pull request and now no more truncation. I need to go back though and look at what I was testing last week because I can't remember why i got rid of what you sent me originally. Maybe I thought that the pull request was already merged and that's why I took it from master? I'm going to check as well to see if I get that issue again with duplicating proxy.php? in the request URL. It seemed to only happen if trying to access tiled services. Weird.

Anyway, truncation gone. Thanks for the reminder. Now to test a couple of other things:-) Thanks for the help Doug:-) @phpmaps
